File: JULY-2025/11-07-2025/p2402.java
class Solution {
    public int mostBooked(int n, int[][] meetings) {
        Arrays.sort(meetings, (a, b) -> Integer.compare(a[0], b[0]));


        PriorityQueue<Integer> availableRooms = new PriorityQueue<>();
        for (int i = 0; i < n; i++) {
            availableRooms.offer(i);
        }

        
        PriorityQueue<long[]> occupiedRooms = new PriorityQueue<>((a, b) -> {
            if (a[0] == b[0]) return Long.compare(a[1], b[1]);
            return Long.compare(a[0], b[0]);
        });

        
        int[] meetingCounts = new int[n];

        for (int[] meeting : meetings) {
            int start = meeting[0];
            int end = meeting[1];

            
            while (!occupiedRooms.isEmpty() && occupiedRooms.peek()[0] <= start) {
                availableRooms.offer((int) occupiedRooms.poll()[1]);
            }

            if (!availableRooms.isEmpty()) {
               
                int room = availableRooms.poll();
                occupiedRooms.offer(new long[]{end, room});
                meetingCounts[room]++;
            } else {
               
                long[] earliest = occupiedRooms.poll();
                long newEnd = earliest[0] + (end - start);
                int room = (int) earliest[1];
                occupiedRooms.offer(new long[]{newEnd, room});
                meetingCounts[room]++;
            }
        }

        
        int maxMeetings = 0, resultRoom = 0;
        for (int i = 0; i < n; i++) {
            if (meetingCounts[i] > maxMeetings) {
                maxMeetings = meetingCounts[i];
                resultRoom = i;
            }
        }

        return resultRoom;

        
    }
}

File: JULY-2025/11-07-2025/p2402.py
class Solution:
    def mostBooked(self, n: int, meetings: List[List[int]]) -> int:
        meetings.sort()

        available_rooms = list(range(n))
        heapq.heapify(available_rooms)

        busy_rooms = []

        room_meeting_count = [0] * n

        for start, end in meetings:
            while busy_rooms and busy_rooms[0][0] <= start:
                free_time, room = heapq.heappop(busy_rooms)
                heapq.heappush(available_rooms, room)

            if available_rooms:
                room = heapq.heappop(available_rooms)
                heapq.heappush(busy_rooms, (end, room))
            else:
                earliest_end_time, room = heapq.heappop(busy_rooms)
                duration = end - start
                new_end = earliest_end_time + duration
                heapq.heappush(busy_rooms, (new_end, room))
        
            room_meeting_count[room] += 1

        max_count = max(room_meeting_count)
        for i in range(n):
            if room_meeting_count[i] == max_count:
                return i 

        